backends/p4tools/modules/testgen/options.h

```diff
@@ -65,6 +65,14 @@ class TestgenOptions : public AbstractP4cToolOptions {
     /// for statement reachability analysis.
     bool dcg = false;
 
+    /// Add conditions defined in assert/assume to the path conditions.
+    /// Only tests which satisfy these conditions can be generated.
+    bool enforceAssumptions = false;
```

Comment:
Do you plan on there being a way to only add conditions for `assume`, but not for `assert`?

I could imagine one might want to enforce `assume` conditions in tests created, but at the same time try to find tests that violate `assert` conditions.

Why?  `assume` would be conditions of "I am telling you I will be operating in an environment where these conditions are true, for reasons that are not necessarily possible to infer from the P4 source code, e.g. because I know I am operating in an environment where I will never see these protocol/NextHeader values.", but `assert` statements would be conditions of "I am hoping these conditions are always true, but want the tools to catch me if I am wrong in this hope."

Reply:
Yes, this is still in draft mode. Currently, we have two flags. 

- If only `enforceAssumptions` is active, P4Testgen will only generate tests that satisfy all assert and assume conditions.
- If also `genAssertionViolations` is active, P4Testgen will try to only produce tests that violate assertions also including assumptions in assume calls.
- If only  `genAssertionViolations` is active, P4Testgen will not add assumptions to the path and only generate tests that violate asserts. 

The tricky part is generating only violating tests while exploring the full program(i.e., not stopping at the first violating assert).

Reply:
This feature is now ready and can be co-developed with https://github.com/p4lang/p4-spec/issues/1207. Testgen can implement the specified functionality with assumption and assertion mode. The question is how to write tests here.